usr/src/uts/common/fs/zfs/dsl_dataset.c

```diff
@@ -1971,7 +1972,24 @@ dsl_dataset_space(dsl_dataset_t *ds,
     uint64_t *refdbytesp, uint64_t *availbytesp,
     uint64_t *usedobjsp, uint64_t *availobjsp)
 {
-	*refdbytesp = dsl_dataset_phys(ds)->ds_referenced_bytes;
+	dsl_dir_t *dd = ds->ds_dir;
+	objset_t *os = dd->dd_pool->dp_meta_objset;
+
+	if (dsl_dir_is_zapified(dd) &&
+		zap_contains(os, dd->dd_object, DD_FIELD_COMPRESS_QUOTA) == 0)
+	{
+		uint64_t compress_quota;
+		VERIFY0(zap_lookup(os, dd->dd_object, DD_FIELD_COMPRESS_QUOTA, 8, 1, &compress_quota));
+		if (compress_quota)
+			*refdbytesp = dsl_dataset_phys(ds)->ds_referenced_bytes
+				+ dsl_dataset_phys(ds)->ds_uncompressed_bytes
+				- dsl_dataset_phys(ds)->ds_compressed_bytes;
```

Comment:
When there is no compression, there is a delta between used_bytes and logical_used bytes.
To be honnest, I didn't found the exact explanation in code, just some suposition.
Maybe you can enlighten me ? :)

To retreive this exact value, and make the behaviour the same for the same data written, I compute this delta.
Best way to see that is throught experimenting. I copy the exact same file, changing compression : 

NAME	         COMPRESS  COMPRESS_QUOTA   USED        LUSED     REFER    LREFER
test/a	         off             	    off   				                          4485120   4470784   4485120   4470784
test/aa	       on             	    off   		                          		1399808   4470784   1399808   4470784
test/aaa     	on                 on   	                          		1399808   4470784   4485120   4470784

The delta here is :  4485120 - 4470784 = 14336

I do the same in dsl_dir.